src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -968,21 +968,40 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
                             _rcwCache.Remove(identity);
                         }
                     }
+                }
+
+                if (wrapperMaybe is not null)
+                {
+                    retValue = wrapperMaybe;
 
-                    if (wrapperMaybe is not null)
+                    using (_lock.EnterScope())
                     {
-                        retValue = wrapperMaybe;
-                        NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
-                            identity,
-                            inner,
-                            this,
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
+                        if (!s_rcwTable.TryGetValue(retValue, out NativeObjectWrapper? wrapper))
                         {
-                            wrapper.Release();
-                            throw new NotSupportedException();
+                            // Exit the lock before we create the native object wrapper
+                            // as we may execute arbitrary user code.
+                            _lock.Exit();
+                            try
+                            {
+                                NativeObjectWrapper newWrapper = NativeObjectWrapper.Create(
+                                    identity,
+                                    inner,
+                                    this,
+                                    retValue,
+                                    flags);
+
+                                wrapper = s_rcwTable.GetValue(retValue, _ => newWrapper);
+                                if (wrapper != newWrapper)
+                                {
+                                    newWrapper.Release();
+                                }
+                            }
+                            finally
+                            {
+                                _lock.Enter();
+                            }
                         }
+
                         _rcwCache.Add(identity, wrapper._proxyHandle);
                         AddWrapperToReferenceTrackerHandleCache(wrapper);
```

Comment:
Trying to understand this better, why is this call needed if we lost the race? In that case, someone else managed to create a new wrapper for the target object. Doesn't that mean that that same someone else would've called `AddWrapperToReferenceTrackerHandleCache` with the same exact wrapper already, meaning doing it here would cause the same wrapper to be registered to the cache multiple times? I can see that cache is basically a custom hash set, meaning it should still work fine, but I'm wondering whether we could get better perf by skipping this call if we lost the race? Or does this not matter because we expect losing the race would be very rare?